modules/core/arrow-annotations-processor/src/main/java/arrow/renzu/RenzuGenerator.kt

```diff
@@ -0,0 +1,181 @@
+package arrow.renzu
+
+import arrow.common.messager.log
+import arrow.common.utils.*
+import arrow.instances.AnnotatedInstance
+import javaslang.Tuple2
+import org.jetbrains.kotlin.serialization.deserialization.TypeTable
+import org.jetbrains.kotlin.serialization.deserialization.supertypes
+import java.io.File
+import javax.lang.model.element.Name
+
+data class Instance(val target: AnnotatedInstance) {
+  val name: Name = target.classElement.simpleName
+  val arrowModule: String = target.classOrPackageProto.`package`.substringAfterLast(".")
+
+  /**
+   * Returns any implemented typeclasses at any level crawling the instance hierarchy.
+   */
+  fun implementedTypeclasses(): List<ClassOrPackageDataWrapper.Class> =
+    target.superTypes.filter { it.fullName.removeBackticks().contains("typeclass") }
+}
+
+data class TypeClass(val processor: RenzuProcessor, val simpleName: String, private val classWrapper: ClassOrPackageDataWrapper.Class) {
+  val arrowModule: String = classWrapper.`package`.substringAfterLast(".")
+
+  override fun equals(other: Any?): Boolean =
+    if (other !is TypeClass) false else simpleName == other.simpleName
+
+  override fun hashCode(): Int = simpleName.hashCode()
+
+  /**
+   * Returns the typeclasses that are a direct parent of this one, for composing the typeclass
+   * relationship on the tree.
+   */
+  fun parentTypeClasses(current: ClassOrPackageDataWrapper.Class): List<TypeClass> {
+    val typeTable = TypeTable(current.classProto.typeTable)
+    val interfaces = current.classProto.supertypes(typeTable).map {
+      it.extractFullName(current)
+    }.filter {
+      it != "`kotlin`.`Any`"
+    }
+    val parentInterfaces: List<ClassOrPackageDataWrapper.Class> = listOf()
+    return when {
+      interfaces.isEmpty() -> parentInterfaces.map {
+        TypeClass(processor, it.simpleName, it)
+      }
+      else -> {
+        interfaces.flatMap { i ->
+          try {
+            val className = i.removeBackticks().substringBefore("<")
+            val typeClassElement = processor.elementUtils.getTypeElement(className)
+            val parentInterface = processor.getClassOrPackageDataWrapper(typeClassElement) as ClassOrPackageDataWrapper.Class
+            if (i.removeBackticks().contains("typeclass")) {
```

Comment:
Can this fail if the type class extends an interface in the same package that is not a type class?